stream-parent/stream-client/src/main/java/org/interledger/stream/sender/AimdCongestionController.java

```diff
@@ -0,0 +1,309 @@
+   *                           sizes increase as the payment path continues to process larger and larger packets.
+   * @param decreaseFactor     An {@link UnsignedLong} representing the amount to lower {@link #maxInFlight} by when the
+   *                           receiver rejects a STREAM packet using a T04 or non-F08 error code (F08 rejections will
+   *                           contain information that can be used to reduce
+   * @param streamCodecContext A {@link CodecContext} for encoding and decoding STREAM packets and frames.
+   */
+  public AimdCongestionController(
+      final UnsignedLong startAmount, final UnsignedLong increaseAmount, final BigDecimal decreaseFactor,
+      final CodecContext streamCodecContext
+  ) {
+    this.maxInFlight = new AtomicReference<>(Objects.requireNonNull(startAmount, "startAmount must not be null"));
+    this.increaseAmount = Objects.requireNonNull(increaseAmount, "increaseAmount must not be null");
+    this.decreaseFactor = Objects.requireNonNull(decreaseFactor, "decreaseFactor must not be null");
+    this.streamCodecContext = Objects.requireNonNull(streamCodecContext, "streamCodecContext must not be null");
+
+    this.congestionState = new AtomicReference<>(CongestionState.SLOW_START);
+    this.amountInFlight = new AtomicReference<>(UnsignedLong.ZERO);
+
+    this.maxPacketAmount = Optional.empty();
+  }
+
+  /**
+   * <p>Compute the maximum packet amount that should be used if a new ILPv4 packet is going to be sent as part of this
+   * Stream. This value is used by the stream client to form the next prepare packet size, and fluctuates based upon
+   * prior stream activity.</p>
+   *
+   * <p>This computation depends on a sub-computation called `amountLeftInWindow`, which is the difference between the
+   * {@link #maxInFlight} and the current {@link #amountInFlight}. The {@code maxAmount} returned by this function is
+   * then computed by taking the min of `amountLeftInWindow` and {@link #maxPacketAmount}.</p>
+   *
+   * @return An {@link UnsignedLong} representing the current max packet amount for packets in this stream.
+   */
+  @Override
+  public UnsignedLong getMaxAmount() {
+    // A "window" is just an amount of time where some value is in-flight, and a StreamClient wants to put more
+    // in-flight. Thus, the amount left in a window will trend towards zero as the amountInFlight increases.
+    final UnsignedLong amountLeftInWindow = maxInFlight.get().minus(amountInFlight.get());
+
+    // Synchronization is _probably not_ needed here since we're not assigning any values that could get corrupted, and
+    // usage of this return value should be considered as a snapshot at a given moment of time anyway...
+    return this.maxPacketAmount
+        // If maxInFlight is specified, take the min of `amountLeftInWindow` and `maxInFlight`.
+        .map(maxPacketAmount -> StreamUtils.min(amountLeftInWindow, maxPacketAmount))
+        .orElse(amountLeftInWindow);
```

Comment:
Whenever I see
> Synchronization is _probably not_ needed here
I immediately read the code probably 10 times.

I don't understand enough about STREAM yet to really speak intelligently, but the answer I have so far is you're _probably_ right. As far as I can tell, the worst case would be that, for some amount of time, you're allowing a larger amount than you actually should, but I'd assume that it's within reason that this isn't expecting to be immediately consistent on those changes? Otherwise it seems like you could create a severe bottleneck trying to block.

All of that aside, the comment says 
> // If maxInFlight is specified, take the min of `amountLeftInWindow` and `maxInFlight`. 

but we're comparing `StreamUtils.min(amountLeftInWindow, maxPacketAmount)`, so one of those is incorrect.